pkg/login/auth_test.go

```diff
@@ -4,177 +4,157 @@ import (
 	"errors"
 	"testing"
 
-	. "github.com/smartystreets/goconvey/convey"
-
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/services/ldap"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestAuthenticateUser(t *testing.T) {
-	Convey("Authenticate user", t, func() {
-		authScenario("When a user authenticates without setting a password", func(sc *authScenarioContext) {
-			mockLoginAttemptValidation(nil, sc)
-			mockLoginUsingGrafanaDB(nil, sc)
-			mockLoginUsingLDAP(false, nil, sc)
-
-			loginQuery := models.LoginUserQuery{
-				Username: "user",
-				Password: "",
-			}
-			err := authenticateUser(&loginQuery)
-
-			Convey("login should fail", func() {
-				So(sc.grafanaLoginWasCalled, ShouldBeFalse)
-				So(sc.ldapLoginWasCalled, ShouldBeFalse)
-				So(err, ShouldEqual, ErrPasswordEmpty)
-				So(sc.loginUserQuery.AuthModule, ShouldEqual, "")
-			})
-		})
+	authScenario(t, "When a user authenticates without setting a password", func(sc *authScenarioContext) {
+		mockLoginAttemptValidation(nil, sc)
+		mockLoginUsingGrafanaDB(nil, sc)
+		mockLoginUsingLDAP(false, nil, sc)
+
+		loginQuery := models.LoginUserQuery{
+			Username: "user",
+			Password: "",
+		}
+		err := authenticateUser(&loginQuery)
 
-		authScenario("When a user authenticates having too many login attempts", func(sc *authScenarioContext) {
-			mockLoginAttemptValidation(ErrTooManyLoginAttempts, sc)
-			mockLoginUsingGrafanaDB(nil, sc)
-			mockLoginUsingLDAP(true, nil, sc)
-			mockSaveInvalidLoginAttempt(sc)
-
-			err := authenticateUser(sc.loginUserQuery)
-
-			Convey("it should result in", func() {
-				So(err, ShouldEqual, ErrTooManyLoginAttempts)
-				So(sc.loginAttemptValidationWasCalled, ShouldBeTrue)
-				So(sc.grafanaLoginWasCalled, ShouldBeFalse)
-				So(sc.ldapLoginWasCalled, ShouldBeFalse)
-				So(sc.saveInvalidLoginAttemptWasCalled, ShouldBeFalse)
-				So(sc.loginUserQuery.AuthModule, ShouldEqual, "")
-			})
-		})
+		require.EqualError(t, err, ErrPasswordEmpty.Error())
+		assert.False(t, sc.grafanaLoginWasCalled)
+		assert.False(t, sc.ldapLoginWasCalled)
+		assert.Equal(t, "", sc.loginUserQuery.AuthModule)
```

Comment:
```suggestion
		assert.Empty(t, sc.loginUserQuery.AuthModule)
```

Reply:
Yeah, I see what you mean. I thought about it too, and I actually think it makes sense to explicitly compare the value to a string literal here.

The rationale being if for example the type of `AuthModule` changes from `string` to `*string`, then the test would pass if the value is either `nil` or `""`, while we might want it to fail if it's `nil`.

What do you think?

Reply:
I guess you're making a valid point, but if we were to follow this logic we would have to generally avoid using `assert.Empty` to test empty strings. I also don't see any reason why `LoginUserQuery` should become a pointer, generally string fields shouldn't be pointers in Go. Their absence should be represented by the empty string rather than `nil`.

Reply:
> I also don't see any reason why `LoginUserQuery` should become a pointer, since generally string fields shouldn't be pointers in Go

:+1: I agree. I updated to `assert.Empty` since it's what the project seems to prefer.

(Updated my rewrite rules too for this case too. :smile: )